pinot-segment-local/src/main/java/org/apache/pinot/segment/local/dedup/ConcurrentMapPartitionDedupMetadataManager.java

```diff
@@ -19,106 +19,124 @@
 
+class ConcurrentMapPartitionDedupMetadataManager extends BasePartitionDedupMetadataManager {
   @VisibleForTesting
-  final ConcurrentHashMap<Object, IndexSegment> _primaryKeyToSegmentMap = new ConcurrentHashMap<>();
+  final AtomicDouble _largestSeenTime = new AtomicDouble(0);
+  @VisibleForTesting
+  final ConcurrentHashMap<Object, Pair<IndexSegment, Double>> _primaryKeyToSegmentAndTimeMap =
+      new ConcurrentHashMap<>();
 
-  public ConcurrentMapPartitionDedupMetadataManager(String tableNameWithType, List<String> primaryKeyColumns,
-      int partitionId, ServerMetrics serverMetrics, HashFunction hashFunction) {
-    _tableNameWithType = tableNameWithType;
-    _primaryKeyColumns = primaryKeyColumns;
-    _partitionId = partitionId;
-    _serverMetrics = serverMetrics;
-    _hashFunction = hashFunction;
+  protected ConcurrentMapPartitionDedupMetadataManager(String tableNameWithType, int partitionId,
+      DedupContext dedupContext) {
+    super(tableNameWithType, partitionId, dedupContext);
   }
 
-  public void addSegment(IndexSegment segment) {
-    // Add all PKs to _primaryKeyToSegmentMap
-    Iterator<PrimaryKey> primaryKeyIterator = getPrimaryKeyIterator(segment);
-    while (primaryKeyIterator.hasNext()) {
-      PrimaryKey pk = primaryKeyIterator.next();
-      _primaryKeyToSegmentMap.put(HashUtils.hashPrimaryKey(pk, _hashFunction), segment);
+  @Override
+  protected void doAddOrReplaceSegment(IndexSegment oldSegment, IndexSegment newSegment,
+      Iterator<DedupRecordInfo> dedupRecordInfoIteratorOfNewSegment) {
+    String segmentName = newSegment.getSegmentName();
+    while (dedupRecordInfoIteratorOfNewSegment.hasNext()) {
+      DedupRecordInfo dedupRecordInfo = dedupRecordInfoIteratorOfNewSegment.next();
+      double dedupTime = dedupRecordInfo.getDedupTime();
+      _largestSeenTime.getAndUpdate(time -> Math.max(time, dedupTime));
+      _primaryKeyToSegmentAndTimeMap.compute(HashUtils.hashPrimaryKey(dedupRecordInfo.getPrimaryKey(), _hashFunction),
+          (primaryKey, segmentAndTime) -> {
+            if (segmentAndTime == null) {
+              return Pair.of(newSegment, dedupTime);
+            } else {
+              // when oldSegment is null, it means we are adding a new segment
+              // when oldSegment is not null, it means we are replacing an existing segment
+              if (oldSegment == null) {
+                _logger.warn("When adding a new segment: dedup record in segment: {} with primary key: {} and dedup "
+                        + "time: {} already exists in segment: {} with dedup time: {}", segmentName,
+                    dedupRecordInfo.getPrimaryKey(), dedupTime, segmentAndTime.getLeft().getSegmentName(),
+                    segmentAndTime.getRight());
+              } else {
+                if (segmentAndTime.getLeft() != oldSegment) {
+                  _logger.warn("When replacing a segment: dedup record in segment: {} with primary key: {} and dedup "
```

Comment:
would this happen? 

as the get-check-update logic is done within ConcurrentMap.compute() and it's atomic, no?

Reply:
It may not happen, just log here in case there are corner cases that we are not aware of. For example, for some reason, a primary key is overriden by another record before the segment replacement logic takes place.